### ISSUE TITLE
Can't get console on running VM OpenShift

### ISSUE THREAD

**Issue by jeremyrogers** [kubevirt/kubevirt#2173]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind enhancement

**What happened**:
I stood up kubevirt on a small OpenShift cluster, and applied a VirtualMachine and then started it with `virtctl start cirros` (manifests are down below).

When running `virtctl console cirros`, I get the output:
```
Successfully connected to cirros console. The escape sequence is ^]
```

Then my shell hangs, and I can only hit the escape sequence to get out. In the API logs, there is this log from the kubevirt API pod:

```
ERROR: logging before flag.Parse: W0404 17:06:23.687833 1 reflector.go:270] pkg/virt-api/api.go:1014: watch of *v1.VirtualMachineInstancePreset ended with: The resourceVersion for the provided watch is too old.
{"component":"virt-api","contentLength":325,"level":"info","method":"GET","pos":"filter.go:46","proto":"HTTP/2.0","remoteAddress":"10.34.1.1","statusCode":200,"timestamp":"2019-04-04T17:06:25.951208Z","url":"/apis/subresources.kubevirt.io/v1alpha3/healthz","username":"-"}
{"component":"virt-api","kind":"VirtualMachineInstance","level":"info","msg":"Websocket connection upgraded","name":"cirros","namespace":"kubevirt","pos":"subresource.go:105","timestamp":"2019-04-04T17:06:30.221206Z","uid":"fa858108-5646-11e9-8514-005056bbfd66"}
ERROR: logging before flag.Parse: W0404 17:06:30.221286 1 client_config.go:548] Neither --kubeconfig nor --master was specified. Using the inClusterConfig. This might not work.
{"component":"virt-api","kind":"VirtualMachineInstance","level":"error","msg":"failed to exectue command [/usr/share/kubevirt/virt-launcher/sock-connector /var/run/kubevirt-private/fa858108-5646-11e9-8514-005056bbfd66/virt-serial0] on the pod connection failed: command terminated with exit code 143","name":"cirros","namespace":"kubevirt","pos":"subresource.go:115","timestamp":"2019-04-04T17:06:30.329247Z","uid":"fa858108-5646-11e9-8514-005056bbfd66"}
```

`oc get vmis` is reporting the VM as Running, and I can get a shell on the VM by running `oc exec -it <pod_name> -c compute /bin/bash`.

While debugging I ran `/usr/share/kubevirt/virt-launcher/sock-connector /var/run/kubevirt-private/fa858108-5646-11e9-8514-005056bbfd66/virt-serial0` in the VM and saw similar behavior, my TTY hung, and that log showed up again.

**What you expected to happen**:

Console access to the VM

**How to reproduce it (as minimally and precisely as possible)**:

I'm not entirely sure, this seems like it could be a configuration issue. 
Install kubevirt 0.15 with `oc apply -f https://github.com/kubevirt/kubevirt/releases/download/v0.15.0/kubevirt.yaml`. I also added the privileged SCC to the three service accounts created in the `kubevirt` namespace : `kubevirt-apiserver`, `kubevirt-controller`, and `kubevirt-handler`.

**Environment**:
- KubeVirt version (use `virtctl version`):
```
Client Version: version.Info{GitVersion:"v0.15.0", GitCommit:"432f91f3a2bdb3c2624377ad327270c9e146d845", GitTreeState:"clean", BuildDate:"2019-03-05T08:21:47Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{GitVersion:"v0.15.0", GitCommit:"432f91f3a2bdb3c2624377ad327270c9e146d845", GitTreeState:"clean", BuildDate:"2019-03-05T09:55:30Z", GoVersion:"go1.11.4", Compiler:"gc", Platform:"linux/amd64"}
```

- Kubernetes version (use `kubectl version`):
```
oc v3.11.0+0cbc58b
kubernetes v1.11.0+d4cacc0
features: Basic-Auth

Server https://<hostname>:443
openshift v3.11.0+06cfa24-67
kubernetes v1.11.0+d4cacc0
```

- VM or VMI specifications:
```
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  creationTimestamp: 2019-04-03T19:28:48Z
  generation: 1
  name: cirros
  namespace: kubevirt
  resourceVersion: "24546251"
  selfLink: /apis/kubevirt.io/v1alpha3/namespaces/kubevirt/virtualmachines/cirros
  uid: b467bc8d-5646-11e9-8514-005056bbfd66
spec:
  running: true
  template:
    metadata:
      creationTimestamp: null
      labels:
        kubevirt.io/domain: testvm
        kubevirt.io/size: small
    spec:
      domain:
        devices:
          disks:
          - disk:
              bus: virtio
            name: containerdisk
          - disk:
              bus: virtio
            name: cloudinitdisk
          interfaces:
          - bridge: {}
            name: default
        machine:
          type: ""
        resources:
          requests:
            memory: 64M
      networks:
      - name: default
        pod: {}
      volumes:
      - containerDisk:
          image: kubevirt/cirros-registry-disk-demo
        name: containerdisk
      - cloudInitNoCloud:
          userDataBase64: SGkuXG4=
        name: cloudinitdisk
status:
  created: true
  ready: true
```

- Cloud provider or hardware configuration:
VMs on VMware hosts
- OS (e.g. from /etc/os-release):
```
NAME="Red Hat Enterprise Linux Server"
VERSION="7.6 (Maipo)"
ID="rhel"
ID_LIKE="fedora"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="7.6"
PRETTY_NAME="Red Hat Enterprise Linux Server 7.6 (Maipo)"
```

- Kernel (e.g. `uname -a`):
```
Linux<hostname>.ccs.ornl.gov 3.10.0-957.el7.x86_64 #1 SMP Thu Oct 4 20:48:51 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by petrkotas** [comment]
Hi @jeremyrogers, 
thanks for reporting this issue.

Can you please also try virsh console in the pod?
```
virsh list
virsh console <vm_name> 
```
this should fail with some error or connect. If some, what is the error?

There is one thing that can cause troubles in your VM. The namespace. You are running vm in KubeVirt private namespace `kubevirt`. 

It will cause issues when you run `virctl console cirros` which will look into namespace `default` and not in the namespace `kubevirt`. To connect properly to the VM, you can use `virctl console --namespace kubevirt cirros`.

**Comment by petrkotas** [comment]
Openshift cluster just spinned up. I have tried your VM and I can connect to console without any issue.
I have moved the VM to default namespace though.

Looking forward for your feedback.

**Comment by jeremyrogers** [comment]
Thanks for the reply. 

Here's the output from the `virsh` commands from inside the pod:

```
[root@cirros /]# virsh list
 Id   Name              State
---------------------------------
 1    kubevirt_cirros   running

[root@cirros /]# virsh console kubevirt_cirros
Connected to domain kubevirt_cirros
Escape character is ^]
error: internal error: character device console0 is not using a PTY
```

I tried `virtctl console --namespace kubevirt cirros`, but saw the same behavior. I don't think that's a namespacing issue. it seems like virtctl uses my current kube context, since if I switch my context's namespace (such as running `oc project default`), and then run `virtctl console cirros`, I get the error `Can't connect to websocket (404): VirtualMachineInstance cirros in namespace default not found.`. When I switch back to the kubevirt namespace (with `oc project kubevirt`), I see the error from above. Let me know if there's more info I can get you. I'm about to try to run a VM on a different OpenShift cluster, so I'll see if the same error appears there.

**Comment by petrkotas** [comment]
@jeremyrogers thank you for your help. Can you please provide logs from apiserver from kube-system, virt-api server (both) from kubevirt namespace and from virt-controller from kubevirt namespace?

Would it be possible for you to disable selinux for one run? I am curious whether it is not blocked by it.

Thank you.

**Comment by jeremyrogers** [comment]
Also, for what it's worth, I downgraded kubevirt and virtctl to 0.14.0 and this did not fix the problem

**Comment by petrkotas** [comment]
This is really weird. The errors you can see in the logs, are not fatal. They are present because of connection switching. They should not be there, but are not responsible for hanging console.

The issue seems to be in the VM, since it hangs even on the direct socket connection via `sock-connector`. 

What is really weird, is that the issue is persistent across different set ups. Sorry for not asking right away. What is the cluster state (oc get pods --all-namespaces)? Also would you please try different VM, e.g. (`kubevirt/fedora-cloud-container-disk-demo`)? 

If the problem is persistent please post all three: VM, VMI and POD yaml files from the cluster. 

Thanks you for your help!

**Comment by rmohr** [comment]
@jeremyrogers did you try to access the VMI with `virtctl vnc myvmi -n mynamespace`? Very often the issue is that the VMI is not booting. Then you simply don't see anything on the serial console. It then looks like the serial console hangs. In that case you normally see something like `no bootable device found` on the vnc screen.

**Comment by jeremyrogers** [comment]
@rmohr That's a good idea, looks like this is the problem. I moved to using the fedora image that @petrkotas suggested, since the cirros image was working sometimes. The fedora image seems to consistently fail, which is good for debugging.

I ran `virtctl vnc fedora` and saw this error:
```
Initramfs unpacking failed: junk in compressed archive
```

Since that led me to think it was a memory issue, I did notice that the fedora VM was using the default memory limit, which was set to 64M, which seems a little low for a full VM. I raised that to 1Gi request/2Gi limit, and now I was able to spin the VM up successfully. Thanks for your help guys, I think this issue can be considered closed now, I'll reopen if I need.
